### PR TITLE
Add a convert_to_gpkg function to the layer source object

### PR DESCRIPTION
**Pull request by nirvn**
This new function is used in the new qfieldsync "convert to cloud project" feature (more bits coming in on the qfieldsync repository).

The function takes a source layer and writes a geopackage equivalent in the provided destination directory. If the source layer is a geopackage, the function merely copies the dataset.

If the source layer isn't a geopackage, it'll make a copy of the original dataset while being careful to respect pre-existing subset string filter, and insuring that joined fields as well as virtual fields remain as such (i.e. aren't made permanent).

### CODE REVIEW

File: layer.py [hunk 1]
```diff
@@ -417,7 +421,105 @@ def copy(self, target_path, copied_files, keep_existent=False):
...
+                    file_path = parts["path"]
+        if file_path == "":
+            file_path = uri_parts[0]
+
+        dest_file = ""
+        new_source = ""
+        if os.path.isfile(file_path):
+            # check if the source is a geopackage, and merely copy if it's the case
+            source_path, file_name = os.path.split(file_path)
+            if file_name.endswith(".gpkg"):
```
Comment (suricactus) [inline layer.py:461]: Any chance to use something like:
```
QgsVectorDataProvider::storageType()
```


<br class="Apple-interchange-newline">

Comment (nirvn) [reply layer.py:461]: @suricactus , completely forgot about this one, good call.

**Comment by nirvn** [comment]
@suricactus , comment addressed.

File: layer.py [hunk 2]
```diff
@@ -417,7 +421,106 @@ def copy(self, target_path, copied_files, keep_existent=False):
...
+        path.
+
+        :param layer: The layer to copy
+        :param target_path: A path to a folder into which the data will be copied
+        :param keep_existent: if True and target file already exists, keep it as it is
+        """
+
+        if (
+            not self.layer.type() == QgsMapLayer.VectorLayer
+            or not self.layer.dataProvider()
```
Comment (suricactus) [inline layer.py:437]:
```suggestion
            or not self.layer.isValid()
```
Can we change this and continue.

**Review by suricactus — APPROVED** [review]

**Comment by nirvn** [comment]
@suricactus , yep, thanks for the review.

**Comment by suricactus** [comment]
Sorry, hijacked your PR by accident. Too hard to revert since I did a merge commit from master. I will take care.

**Pull request merged**